### ISSUE TITLE
Add `encode_gzip`/`decode_gzip` functions to VRL

### ISSUE THREAD

**Issue by arch-xtof** [vectordotdev/vector#15978]
### A note for the community

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Use Cases

I have GELF UDP messages that are compressed using GZIP, unfortunately, there is no option to disable this compression. It would be nice to have a feature to decompress the input using VRL

### Attempted Solutions

Use Lua

### Proposal

have `encode_gzip`/`decode_gzip` functions in VRL

### References

https://github.com/vectordotdev/vector/issues/10421 Similar issue for zlib

### Version

0.26.0

**Comment by zamazan4ik** [comment]
@jszwedko since we have a similar request but for another algorithm, we need to discuss the interface details here.

The first question: do we want to have a different function `encode_X\decode_X` for every requested compression algorithm? I suppose - yes, as it's already done in the 3rd party libraries since different algorithms have different parameters (e.g. different supported compression levels if we are talking about `encode_X` functions).

The second question: do we want to support specifying compression levels in the `encode_X` functions? From the one hand, it's quite useful for the users to have the ability to specify compression levels since it allows them to control "speed/result size" tradeoff. From another hand, the feature is not requests right now (but adding it later could bring questions regarding the interface compatibility between old `encode_X` without compression level support and the newer one with the support of this feature.

What do you think? From the implementation perspective, I think the requested features will not be difficult to implement since they would be just "simple" VRL wrappers over 3rd party actual implementations (like `flate2`, `async_compression`, and possibly others).

One more point: I think would be a good idea keep in mind implementing the corresponding `encode/decode` pair for ZSTD compression algorithm, since it gains huge adoption over compression world nowadays.

**Comment by zamazan4ik** [comment]
And one more unrelated to the topic of this issue.

I think the issues like this could be marked with a "Good first issue" label. I guess it would help to the new contributors start with something more lightweight from the Vector perspective.

**Comment by jszwedko** [comment]
> @jszwedko since we have a similar request but for another algorithm, we need to discuss the interface details here.
> 
> The first question: do we want to have a different function `encode_X\decode_X` for every requested compression algorithm? I suppose - yes, as it's already done in the 3rd party libraries since different algorithms have different parameters (e.g. different supported compression levels if we are talking about `encode_X` functions).

Yeah, I think separate makes sense for that reason: they can have different parameters.

> The second question: do we want to support specifying compression levels in the `encode_X` functions? On the one hand, it's quite useful for the users to have the ability to specify compression levels since it allows them to control "speed/result size" tradeoff. On another hand, the feature is not requested right now (but adding it later could bring questions regarding the interface compatibility between the old `encode_X` without compression level support and the newer one with the support of this feature.

I'd be fine leaving specifying levels as an enhancement since it would be backwards compatible to add later.

> What do you think? From the implementation perspective, I think the requested features will not be difficult to implement since they would be just "simple" VRL wrappers over 3rd party actual implementations (like `flate2`, `async_compression`, and possibly others).

Agreed, I think this one should be relatively straightforward to implement with the available libraries.

> One more point: I think would be a good idea to keep in mind implementing the corresponding `encode/decode` pair for ZSTD compression algorithm, since it gains huge adoption in the compression world nowadays.

Agreed, zstd would be useful to support too.

Thanks for the questions!
